### PR TITLE
HPCC-30304 Introduce ISecret and simplify the secret calling code

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@afishbeck untested.  For initial discussion.

File: system/jlib/jsecrets.cpp [hunk 1]
```diff
@@ -1029,6 +1046,130 @@ extern jlib_decl bool getSecretValue(StringBuffer & result, const char *category
...
+}
+
+void CSecret::checkStale() const
+{
+    if (isStale())
+    {
+        //MORE: This could block or fail - in roxie especially it would be better to return the old value
+        try
+        {
+            secret.setown(getSecretTree(category, name, vaultId, version));
```
Comment (afishbeck) [inline system/jlib/jsecrets.cpp:1112]: @ghalliday I wonder if getSecretTree() itself shouldn't fail if the cache can't be updated?  If cache expiration itself is basically just a way of keeping the secret fresh.

It's different then say the certificate itself expiring which is separate and tied to the content becoming invalid.

**Review by afishbeck — COMMENTED** [review]
@ghalliday the code looks cleaner, and the new class looks great.  Only thing is I wonder if we should push the "don't fail if cache expired and you can't get to the vault" even further back into the existing API?

Also I think I did a bit of just that don't fail part in one of my outstanding PRs, but I'll rebase when this is merged and to sync up.

**Comment by ghalliday** [comment]
I agree about moving the expiry and removal from the hash table.  I suspect many of the calls won't switch over to using this form, and we still want them to continue if the vault is down.

Also, when a certificate is updated to a file does the filename change?  I suspect not (especially for bare-metal) and I need to add code to look for entries that look like filenames, and has the timestamp of the file if it exists.

**Pull request merged**